platform/engineering/design-docs/2021-05-18-forms-library.md

```diff
@@ -0,0 +1,852 @@
+The `Router` is at the heart of the routing sub-module. It will hold the routes
+and components to render at each route.
+
+**Usage requirements:**
+_None_
+
+**Responsibilities:**
+1. Provide a React context with the form state
+    - `Page` will use this as the page's
+      [`initialValues`](https://formik.org/docs/api/formik#initialvalues-values)
+      - This is because each page is a separate Formik form, so the `Router` is
+        responsible for stitching all these individual forms together into one
+        cohesive experience
+1. Build a list of routes for the form flow from `Chapter` and `Page`
+   components' `path` props
+    - **Important:** `Route`s define valid routes in a form, but those pages
+      live "outside" the normal form flow; these are used for pages such as
+      `/form-saved` and `/error`
+    - `Page`s nested in `Chapter`s will have their paths prefixed with the
+      `Chapter` path
+    - For example:
+      ```jsx
+      <Chapter path="/personal-information">
+        <Page path="/name">...</Page>
+      </Chapter>
+      ```
+      This will result in a path definition of `/personal-information/name`
+1. Re-compute the list of routes before each page navigation (below)
+    - This ensures any data changes that might affect the routes are captured
+1. Provide callbacks in the React context to navigate forward and backward
+   through the list of computed routes
+
+##### `Chapter` Component
+**Purpose:**
+The **optional** `Chapter` component will be used to render the VA.gov "chrome"
+for multi-page forms which have multiple chapters. This "chrome" includes the
+form title, current chapter name, and progress bar.
+
+**Usage requirements:**
+- `Chapter` may only be a child of `Router`
+- Its only children may be `Page`s
+- A `Chapter` must have children
+
+**Responsibilities:**
+- Render the chapter title and progress bar
+- Render the child `Page`s
+
+##### `Page` Component
+**Purpose:**
+A `Page` is used to define the path for a page and render its form contents and
+the normal page chrome.
+
+**Usage requirements:**
+A `Page` may be the child of either `Router` directly or `Chapter`.
+
+**Responsibilities:**
+- If no `overrideFormik` prop is passed
+  - Render a [`<Formik>`](https://formik.org/docs/api/formik) component and pass
+    it some default props for `initialValues`, `validate`, and `onSubmit`
+  - If the save-in-progress React context is available, `Page` will use it to
+    auto-save the form
+- Render the children, passing them the extra props for form state
+  - For an example usage, see [below](#functionality-array-pages)
+  - **Caveat:** This means when the children of `Page` are a collection of
+    `<Field>`s (or similar) instead of the functional component in the example
+    below, they'll all have "extra" props for the form context
+    - This is the simplest way to support both simple and complex form pages,
+      but may be subject to change later if we can figure out a better way
```

Comment:
I'm not sure which "extra" props you're talking about - can you be a bit more specific?

This is also confusing in general, because the example below has a `<Field*>` inside of a `<Page>`, but this text reads as if there is a _different_ use of `<Field*>` inside of `<Page>` where these extra props won't be used.

Reply:
> I'm not sure which "extra" props you're talking about - can you be a bit more specific?

I can call out the two that I know for sure we'll need, but I want to leave it open-ended for adding more in later because I don't know what _all_ we'll need until we get to implementing it.

> This is also confusing in general, because the example below has a `<Field*>` inside of a `<Page>`, but this text reads as if there is a different use of `<Field*>` inside of `<Page>` where these extra props won't be used.

Yeah, that's fair. This section went through a few iterations and I may not have cleaned it up very well. Basically, it goes like this:
1. You can pass children to `<Page>` just like you would for any other component
    ```jsx
    <Page>
      <StringField />
      <NumberField />
    </Page>
    ```
2. But if you needed to know some information about the form state, like, for example, what array page we're on, you can get that data from `<Page>` by passing a custom component that accepts the props describing the form state
    ```jsx
    <Page>
      {({ formData, pathParams }) => ( /* Render your fields using the data from the form state /* )}
    </Page>
    ```

The alternative to this is importing the React Context that `Router` provides and doing something like
```jsx
import { useContext } from 'React';
import { routerContext } from 'formulate`;

export default () => (
  <Page>
    {() => {
      const { formData, pathParams } = useContext(routerContext);
      return (/* Render fields using the form state */);
    }}
  </Page>
);
```
I opted to go with the former approach because I thought it'd be a better developer experience. I'd love to hear from VFS engineers, though!

That said, I'll go back over this and try to clean it up to make it clearer. :+1:

Reply:
I added some more clarity to this section. If it's still confusing, feel free to leave another comment on the new part. :smile: